aptos-move/framework/aptos-framework/sources/simple_delegations.move

```diff
@@ -0,0 +1,844 @@
+        stake::add_stake_with_cap(&owner_cap, delegated_coins);
+
+        // Record the delegation.
+        let pool_address = signer::address_of(&stake_pool_signer);
+        simple_map::add(delegations_map, operator, Delegation {
+            is_active: true,
+            total_stake: amount,
+            pool_address,
+            owner_cap,
+            commission_percentage,
+            debts: vector::empty<CommissionDebt>(),
+        });
+
+        emit_event<DelegateEvent>(
+            &mut delegations.delegate_events,
+            DelegateEvent { operator, voter, pool_address, amount, commission_percentage },
+        );
+    }
+
+    /// Convenient function to allow the delegator to update the voter address in a delegation they made.
+    public entry fun update_voter(delegator: &signer, operator: address, new_voter: address) acquires Delegations {
+        let delegator_address = signer::address_of(delegator);
+        assert_delegation_exists(delegator_address, operator);
+
+        let delegations = borrow_global_mut<Delegations>(delegator_address);
+        let delegations_map = &mut delegations.delegations;
+        let delegation = simple_map::borrow_mut(delegations_map, &operator);
+        let pool_address = delegation.pool_address;
+        let old_voter = stake::get_delegated_voter(pool_address);
+        stake::set_delegated_voter_with_cap(&delegation.owner_cap, new_voter);
+
+        emit_event<UpdateVoterEvent>(
+            &mut delegations.update_voter_events,
+            UpdateVoterEvent { operator, pool_address, old_voter, new_voter },
+        );
+    }
+
+    /// Unlocks commission amount from the stake pool. Operator needs to wait for the amount to become withdrawable
+    /// at the end of the stake pool's lockup period before they can actually can withdraw_commission.
+    ///
+    /// Anyone can call this function as this is an explicit agreement between delegator and operator.
+    public entry fun request_commission(operator: address, delegator: address) acquires Delegations {
+        assert_delegation_exists(delegator, operator);
+
+        let delegations = borrow_global_mut<Delegations>(delegator);
+        let delegations_map = &mut delegations.delegations;
+        let delegation = simple_map::borrow_mut(delegations_map, &operator);
+        assert!(delegation.commission_percentage > 0, error::invalid_argument(EZERO_COMMISSION_RATE));
+        assert!(delegation.is_active, error::invalid_argument(EINACTIVE_DELEGATION));
+
+        // Unlock just the commission portion from the stake pool.
+        let previous_total_stake = delegation.total_stake;
+        let (total_active_stake, accumulated_rewards, commission_amount) = get_accumulated_rewards(delegation);
+        stake::unlock_with_cap(commission_amount, &delegation.owner_cap);
+
+        // Update new total stake amount to reflect the paid commission amount.
+        let updated_total_stake = total_active_stake - commission_amount;
```

Comment:
hmm, I'm not sure I understand the math here. let' say if we stake 1000 coins, have 100 rewards with 10% commission rate.

the first call to request commission will get (1100, 100, 10), and update total_stake to 990 and unlock 10.
the second call to request commission will get (1100, 110, 11) and update total_stake to 979 and unlock another 11?

I probably miss something here

Reply:
If you look at test_end_to_end, it explains how this is done well.

In your example: let' say if we stake 1000 coins, have 100 rewards with 10% commission rate.

If the get_accumulated_rewards returns (1100, 100, 10), that means the total_active_stake is 1100, accumulated rewards is 100 and commission to pay is 10. So after commission (10 is unlocked), total active stake recorded would be 1100 - 10 = 1090. The invariant here is we also track what's the latest amount that BELONGS to the delegator entirely at the last request_commission call (aka all commissions have been settled).

Reply:
oh I miss the part this is updated to total_active - commission instead of previous_active - commission.